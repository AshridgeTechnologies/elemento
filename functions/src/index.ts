import * as functions from 'firebase-functions'
import axios from 'axios'
import express from 'express'

const app = express()

app.use(express.raw({type: 'application/zip'}))
app.use( (req, res, next) => {
    console.log(req.method, req.url)
    next()
})
app.put('/uploadfunctioncontent', async (req, res) => {
    try {
        const data = req.body
        const {path, GoogleAccessId, Expires, Signature} = req.query
        const url = `https://storage.googleapis.com${path}?GoogleAccessId=${GoogleAccessId}&Expires=${Expires}&Signature=${encodeURIComponent(Signature as string)}`
        console.log('upload url', url)

        await axios.put(url, data, {
            headers: {
                'Content-Type': req.headers['content-type'] as string
            }
        })
        res.end()
    } catch (e) {
        console.error(e)
        res.sendStatus(500)
    }
})

export const uploadfunctioncontent = functions.https.onRequest(app)
