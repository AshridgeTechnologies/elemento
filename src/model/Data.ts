import {ElementId, PropertyValue} from './Types'
import Element from './Element'
import BaseElement from './BaseElement'

type Properties = {
    readonly initialValue?: PropertyValue,
    readonly display?: PropertyValue,
}

export default class Data extends BaseElement<Properties> implements Element {
    constructor(
        id: ElementId,
        name: string,
        properties: Properties
    ) {
        super(id, name, properties)
    }

    static is(element: Element): element is Data {
        return element.constructor.name === this.name
    }

    get initialValue() {return this.properties.initialValue}
    get display() {return this.properties.display ?? false}
}