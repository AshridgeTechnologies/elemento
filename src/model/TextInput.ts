import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}