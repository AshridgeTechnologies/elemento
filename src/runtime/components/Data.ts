import {createElement} from 'react'
import {valueLiteral} from '../runtimeFunctions'
import {equals, mergeRight} from 'ramda'

type Properties = {state: any, display?: boolean}

export default function Data({state, display = false}: Properties) {
    const {_path: path, value} = state
    return display ?  createElement('div', {id: path},
        createElement('div', null, path),
        createElement('code', null, valueLiteral(value))) : null
}

Data.State = class State {
    private state: { value?: any | null, updateFn: (changes: object, replace?: boolean)=> void }

    constructor(private props: { value: any | null | undefined }) {
        this.state = {updateFn: () => {throw new Error('updateFn called before injected')}}
    }

    init(updateFn: (changes: object, replace?: boolean)=> void) {
        this.state.updateFn = updateFn  // no effect on external view so no need to update
    }

    setState(changes: {value?: any | null}) {
        const result = new Data.State(this.props)
        result.state = mergeRight(this.state, changes)
        return result
    }

    private updateState(changes: { value?: object, queries?: object }) {
        this.state.updateFn(this.setState(changes), true)
    }

    mergeProps(newState: typeof this) {
        return equals(this.props, newState.props) ? this : new Data.State(newState.props).setState(this.state)
    }

    get value() {
        return this.state.value !== undefined ? this.state.value : this.props.value
    }

    Set(value: any) {
        this.updateState({value})
    }

    Update(changes: object) {
        const newValue = mergeRight(this.value, changes )
        this.updateState({value: newValue})
    }
}

