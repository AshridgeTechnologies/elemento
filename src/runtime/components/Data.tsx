import React from 'react'
import {valueLiteral} from '../runtimeFunctions'

type Properties = {state: any, display?: boolean}

export default function Data({state, display = false}: Properties) {
    const {_path: path, value} = state
    return display ? <div id={path}>
        <div>{path}</div>
        <code>{valueLiteral(value)}</code>
        </div>
         : null
}
