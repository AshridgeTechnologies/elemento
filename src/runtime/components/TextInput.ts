import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {updateState} from '../appData'
import {valueOfProps} from '../runtimeFunctions'

type Properties = {state: {value?: string, _path: string, _controlValue: string | null}, label?: string, maxLength?: number, width?: string | number, multiline?: boolean}

export default function TextInput({state, ...props}: Properties) {
    const {maxLength, label, multiline, width} = valueOfProps(props)
    const inputProps = maxLength !== undefined ? {inputProps: {maxLength}} : {}
    const widthProp = width !== undefined ? {width} : {}
    const sxProps = {sx: {...widthProp}}
    const optionalProps = definedPropertiesOf({label, multiline})
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        updateState(path, {value: updateValue })
    }

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...inputProps,
        ...sxProps,
        ...optionalProps
    })
}
