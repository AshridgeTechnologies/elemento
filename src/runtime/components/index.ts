export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
