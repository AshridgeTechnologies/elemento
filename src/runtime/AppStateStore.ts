import SubscribableStore, {Callback, Id, UnsubscribeFn} from './SubscribableStore'
import {ComponentState} from './components/ComponentState'
import {unique} from 'radash'

export type StoredState = ComponentState<any>

export default class AppStateStore {

    constructor(private store: SubscribableStore = new SubscribableStore()) {
        this.subscribeToAllChanges()
    }

    get(id: string): StoredState {
        return this.store.get(id)
    }

    set<T extends StoredState>(id: string, item: T) {
        item.init(this.appStoreInterface(id), id)
        this.store.set(id, item)
    }

    updateIfChanged<T extends StoredState>(id: string, item: T) {
        const existingItem = this.get(id)
        const latestItem = existingItem ? existingItem.updateFrom(item) : item
        if (latestItem !== existingItem) {
            this.setDeferNotifications(id, latestItem)
        }
    }

    subscribe(id: string, callback: Callback): UnsubscribeFn {
        return this.store.subscribe(id, callback)
    }

    private subscribeToAllChanges() {
        this.store.subscribeAll( (ids: string[]) => this.updateParentItems(ids))
    }

    private updateParentItems(ids: Id[]) {
        const extractParent = (id: Id) => id.replace(/\.\w+$/, '')
        const parentIds = unique(ids.map(extractParent))
        parentIds.forEach( id => this.get(id)?.onChildStateChange() )
    }

    private setDeferNotifications<T extends StoredState>(id: string, item: T) {
        if (!this.store.isNotifyDeferred) {
            this.store.deferNotifications()
            setTimeout(() => this.store.sendNotifications(), 0)
        }
        this.set(id, item)
    }

    private appStoreInterface(id: string) {
        const store = this
        return {
            latest() {
                return store.get(id)
            },
            updateVersion(changes: object) {
                const updatedItem = store.get(id).withMergedState(changes)
                store.set(id, updatedItem)
            },
            getChildState(subPath: string) {
                return store.get(id + '.' + subPath)
            },
            getOrCreateChildState(subPath: string, item: StoredState) {
                store.updateIfChanged(id + '.' + subPath, item)
                return this.getChildState(subPath)
            },
            getApp() {
                const appId = id.split('.')[0]
                return store.get(appId)
            }
        }
    }
}
