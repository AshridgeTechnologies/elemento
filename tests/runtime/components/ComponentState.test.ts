import {expect, test, vi} from "vitest"
/**
 * @vitest-environment jsdom
 */
import {wait} from '../../testutil/testHelpers'
import {appFunctions, globalFunctions, stateProps, wrapFn} from '../../../src/runtime'
import {DataState} from '../../../src/runtime/components/Data'
import {BaseComponentState, Block, Calculation, Data, TextInput} from '../../../src/runtime/components'
import {TextInputState} from '../../../src/runtime/components/TextInput'
import AppStateStore from '../../../src/runtime/AppStateStore'
import SubscribableStore from '../../../src/runtime/SubscribableStore'

const {Log, Left} = globalFunctions
const {Set} = appFunctions

class NameInput_State extends BaseComponentState<any> {
    childNames = ['Block1', 'FirstName', 'SecondName', 'FullName', 'NameData']

    createChildStates() {
        const pathTo = (name: string) => this._path + '.' + name
        const {First, Second, SaveAction} = this.props
        const {SetNames, GetNameAndInitial, SetNameData} = this
        const Block1 = this.getOrCreateChildState('Block1', new Block.State(stateProps(pathTo('Block1')).props))
        const FirstName = this.getOrCreateChildState('FirstName', new TextInput.State(stateProps(pathTo('FirstName')).value(First).props))
        const SecondName = this.getOrCreateChildState('SecondName', new TextInput.State(stateProps(pathTo('SecondName')).value(Second).props))
        const FullName = this.getOrCreateChildState('FullName', new Calculation.State(stateProps(pathTo('FullName')).value(FirstName + ' ' + SecondName).props))
        const NameData = this.getOrCreateChildState('NameData', new Data.State(stateProps(pathTo('NameData')).props))
        return {Block1, FirstName, SecondName, FullName, NameData}
    }

    get Block1() { return this.childStates.Block1 }
    get FirstName() { return this.childStates.FirstName as TextInputState }
    get SecondName() { return this.childStates.SecondName as TextInputState }
    get FullName() { return this.childStates.FullName }
    get NameData() { return this.childStates.NameData as DataState }

    SetNames = wrapFn('NameInput.SetNames', 'calculation', async (firstName, secondName) => {
        const {SetNameData} = this
        await SetNameData(secondName + ', ' + firstName)
    })
    SetNameData = wrapFn('NameInput.SetNameData', 'calculation', (newName) => {
        const {NameData} = this
        const {Initials} = this
        Set(NameData, newName)
        Log('Set Name Data', newName, Initials)
    })
    GetNameAndInitial = wrapFn('NameInput.GetNameAndInitial', 'calculation', () => {
        const {SecondName, FirstName} = this
        return SecondName + ', ' + Left(FirstName, 1)
    })

    Save_action = wrapFn('NameInput.Save', 'action', async () => {
        const {FirstName, SecondName} = this
        const {SaveAction} = this.props
        const {SetNames} = this
        await SaveAction()
        await SetNames(FirstName, SecondName)
    })
    get DisplayName() {
        const {FirstName, SecondName, FullName} = this
        return FirstName + ' ' +  SecondName + ' - aka ' + FullName
    }
    get Initials() {
        const {FirstName, SecondName} = this
        return Left(FirstName, 1)  +   Left(SecondName, 1)
    }
    get NamePlusInitial() {
        const {GetNameAndInitial} = this
        return GetNameAndInitial()
    }
}

test('State class has correct properties and functions', async () => {
    const store = new AppStateStore(new SubscribableStore())
    const saveAction = vi.fn()
    const state = new NameInput_State({SaveAction: saveAction, First: 'Andy', Second: 'Brown' })
    store.set('nameInput1', state)
    await wait()
    expect(state.Initials).toStrictEqual('AB')
    await state.latest().SetNames('Marti', 'Parti')
    expect(state.latest().NameData.value).toBe('Parti, Marti')

    store.set('nameInput1.FirstName', new TextInputState({})._withStateForTest({value: 'Genna'}))
    store.set('nameInput1.SecondName', new TextInputState({})._withStateForTest({value: 'Venna'}))
    await wait()
    expect(state.latest().Initials).toStrictEqual('GV')
})

test('State class updates from an object with new props', () => {
    const saveAction = vi.fn()
    const state1 = new NameInput_State({SaveAction: saveAction, First: 'Andy', Second: 'Brown' })
    const state2 = new NameInput_State({SaveAction: saveAction, First: 'Andy', Second: 'Brown' })
    const state3 = new NameInput_State({SaveAction: saveAction, First: 'Andy', Second: 'Black' })

    const store = new AppStateStore(new SubscribableStore())
    store.set('nameInput1', state1)


    expect(state1.updateFrom(state2)).toBe(state1)
    expect(state1.updateFrom(state3).props.Second).toBe('Black')
})
