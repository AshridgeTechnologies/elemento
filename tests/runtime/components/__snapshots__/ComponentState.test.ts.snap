// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Name Input element produces output containing children 1`] = `
<div
  className="MuiBox-root css-0"
  id="app.page1.nameInput1"
>
  <div
    className="MuiStack-root css-b4hyv-MuiStack-root"
    id="app.page1.nameInput1.Block1"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-55ifni-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1kfabtt-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="app.page1.nameInput1.FirstName"
        id="app.page1.nameInput1.FirstName-label"
      >
        First Name
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="app.page1.nameInput1.FirstName"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-81qg8w"
          >
            <span>
              First Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root css-55ifni-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1kfabtt-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor="app.page1.nameInput1.SecondName"
        id="app.page1.nameInput1.SecondName-label"
      >
        Second Name
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id="app.page1.nameInput1.SecondName"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-81qg8w"
          >
            <span>
              Second Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-disableElevation css-1xp33i0-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    id="app.page1.nameInput1.Save"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Save
  </button>
  <div
    id="app.page1.nameInput1.NameData"
  >
    <div>
      app.page1.nameInput1.NameData
    </div>
    <code>
      undefined
    </code>
  </div>
</div>
`;
